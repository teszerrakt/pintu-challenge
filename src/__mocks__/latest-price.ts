import { IGetLatestPricePayload } from 'src/apis/latestPrice/interface'

export const MOCK_LATEST_PRICE_DATA: IGetLatestPricePayload[] = [
  {
    pair: 'btc/idr',
    priceChanges: {
      latestPrice: '458103826',
      day: '0.66',
      week: '2.26',
      month: '10.47',
      year: '51.33',
    },
    currency: {
      currencySymbol: 'BTC',
      currencyGroup: 'BTC',
      color: '#F78B1A',
      name: 'Bitcoin',
      logo: 'https://s3-ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/circle_BTC.svg',
      listingDate: '2020-09-15T09:43:45Z',
    },
  },
  {
    pair: 'eth/idr',
    priceChanges: {
      latestPrice: '27806628',
      day: '-0.17',
      week: '-1.04',
      month: '-1.83',
      year: '62.84',
    },
    currency: {
      currencySymbol: 'ETH',
      currencyGroup: 'ETH',
      color: '#9011FE',
      name: 'Ethereum',
      logo: 'https://s3-ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/circle_ETH.svg',
      listingDate: '2020-09-15T09:43:46Z',
    },
  },
  {
    pair: 'usdt/idr',
    priceChanges: {
      latestPrice: '15002',
      day: '0.04',
      week: '0.33',
      month: '0.14',
      year: '1.11',
    },
    currency: {
      currencySymbol: 'USDT',
      currencyGroup: 'USDT',
      color: '#15A57C',
      name: 'Tether',
      logo: 'https://s3-ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/circle_USDT.svg',
      listingDate: '2020-09-15T09:43:47Z',
    },
  },
]
