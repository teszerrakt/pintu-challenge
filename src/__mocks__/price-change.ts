import { IGetPriceChangeResponse } from 'src/interface'

export const MOCK_PRICE_CHANGE_DATA: IGetPriceChangeResponse[] = [
  {
    pair: 'busd/idr',
    latestPrice: '15015',
    day: '0.15',
    week: '0.16',
    month: '0.41',
    year: '1.22',
  },
  {
    pair: '1inch/idr',
    latestPrice: '4876',
    day: '-3.86',
    week: '17.98',
    month: '-16.72',
    year: '-54.89',
  },
  {
    pair: 'alpha/idr',
    latestPrice: '1762',
    day: '-4.13',
    week: '3.34',
    month: '11.66',
    year: '-10.19',
  },
  {
    pair: 'ldo/idr',
    latestPrice: '28919',
    day: '-1.28',
    week: '0.00',
    month: '-2.87',
    year: '203.83',
  },
  {
    pair: 'vet/idr',
    latestPrice: '280',
    day: '2.19',
    week: '26.13',
    month: '-2.10',
    year: '-27.84',
  },
]
