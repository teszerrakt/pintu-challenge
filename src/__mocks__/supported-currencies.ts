import { IGetSupportedCurrenciesResponse } from 'src/interface'

export const MOCK_SUPPORTED_CURRENCIES: IGetSupportedCurrenciesResponse[] = [
  {
    currencyGroup: 'BUSD',
    color: '#FEBF11',
    currencySymbol: 'BUSD',
    name: 'Binance USD',
    logo: 'https://s3-ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/circle_BUSD.svg',
    decimal_point: 18,
    listingDate: '2023-02-06T07:00:00Z',
    wallets: [
      {
        currencyGroup: 'BUSD',
        tokenSymbol: 'BUSD',
        decimal_point: 18,
        tokenType: 'ERC-20',
        blockchain: 'Ethereum',
        explorer: 'https://etherscan.io/tx/',
        listingDate: '2023-02-06T07:00:00Z',
        blockchainName: 'Ethereum',
        logo: 'https://s3.ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/blockchain/ERC-20.svg',
      },
      {
        currencyGroup: 'BUSD',
        tokenSymbol: 'BUSDBSC',
        decimal_point: 18,
        tokenType: 'BEP-20',
        blockchain: 'Binance Smartchain',
        explorer: 'https://bscscan.com/tx/',
        listingDate: '2023-02-06T07:00:01Z',
        blockchainName: 'BNB Smart Chain',
        logo: 'https://s3.ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/blockchain/BEP-20+(BSC).svg',
      },
    ],
  },
  {
    currencyGroup: '1INCH',
    color: '#3B76BA',
    currencySymbol: '1INCH',
    name: '1inch',
    logo: 'https://s3-ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/circle_1INCH.svg',
    decimal_point: 18,
    listingDate: '2021-08-26T05:15:18Z',
    wallets: [
      {
        currencyGroup: '1INCH',
        tokenSymbol: '1INCH',
        decimal_point: 18,
        tokenType: 'ERC-20',
        blockchain: 'Ethereum',
        explorer: 'https://etherscan.io/tx/',
        listingDate: '2021-08-26T05:15:18Z',
        blockchainName: 'Ethereum',
        logo: 'https://s3.ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/blockchain/ERC-20.svg',
      },
    ],
  },
  {
    currencyGroup: 'ALPHA',
    color: '#1F66E6',
    currencySymbol: 'ALPHA',
    name: 'Alpha Finance',
    logo: 'https://s3-ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/circle_ALPHA.svg',
    decimal_point: 18,
    listingDate: '2022-01-13T18:15:20Z',
    wallets: [
      {
        currencyGroup: 'ALPHA',
        tokenSymbol: 'ALPHA',
        decimal_point: 18,
        tokenType: 'ERC-20',
        blockchain: 'Ethereum',
        explorer: 'https://etherscan.io/tx/',
        listingDate: '2022-01-13T18:15:20Z',
        blockchainName: 'Ethereum',
        logo: 'https://s3.ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/blockchain/ERC-20.svg',
      },
    ],
  },
  {
    currencyGroup: 'LDO',
    color: '#469BF0',
    currencySymbol: 'LDO',
    name: 'Lido DAO',
    logo: 'https://s3-ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/circle_LDO.svg',
    decimal_point: 18,
    listingDate: '2022-05-20T04:00:18Z',
    wallets: [
      {
        currencyGroup: 'LDO',
        tokenSymbol: 'LDO',
        decimal_point: 18,
        tokenType: 'ERC-20',
        blockchain: 'Ethereum',
        explorer: 'https://etherscan.io/tx/',
        listingDate: '2022-05-20T04:00:18Z',
        blockchainName: 'Ethereum',
        logo: 'https://s3.ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/blockchain/ERC-20.svg',
      },
    ],
  },
  {
    currencyGroup: 'VET',
    color: '#23B2F9',
    currencySymbol: 'VET',
    name: 'VeChain',
    logo: 'https://s3-ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/circle_VET.svg',
    decimal_point: 18,
    listingDate: '2022-07-28T07:00:09Z',
    wallets: [
      {
        currencyGroup: 'VET',
        tokenSymbol: 'VET',
        decimal_point: 18,
        tokenType: 'VeChain',
        blockchain: 'VeChain',
        explorer: 'https://explore.vechain.org/transactions/',
        listingDate: '2022-07-28T07:00:09Z',
        blockchainName: 'VeChain',
        logo: 'https://s3.ap-southeast-1.amazonaws.com/static.pintu.co.id/assets/images/logo/blockchain/Vechain.svg',
      },
    ],
  },
]
