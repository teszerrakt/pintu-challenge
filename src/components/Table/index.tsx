import {
  SortingState,
  createColumnHelper,
  flexRender,
  getCoreRowModel,
  getSortedRowModel,
  useReactTable,
} from '@tanstack/react-table'
import Percentage from 'src/components/Percentage'
import { IGetLatestPricePayload } from 'src/apis/latestPrice/interface'
import { stringToCurrency } from 'src/utils/formatter'
import CryptoLogo from 'src/components/Icons/Logo'
import { useState } from 'react'
import Sorting from '../Icons/Sorting'
import Link from 'next/link'
import { PINTU_BASE_URL } from 'src/constants/env'

export interface ITableProps {
  data: IGetLatestPricePayload[]
}

const columnHelper = createColumnHelper<IGetLatestPricePayload>()

const columns = [
  columnHelper.accessor('currency', {
    header: 'CRYPTO',
    cell: (info) => {
      const value = info.getValue()

      return (
        <Link
          className="flex items-center"
          href={`${PINTU_BASE_URL}/market/${value.currencySymbol}`}
        >
          <CryptoLogo url={value.logo} color={value.color} />
          <div className="flex flex-col w-full ml-6 lg:justify-between lg:flex-row">
            <p className="font-semibold">{value.name}</p>
            <p className="text-slate-400">{value.currencySymbol}</p>
          </div>
        </Link>
      )
    },
    enableSorting: false,
  }),
  columnHelper.accessor('priceChanges.latestPrice', {
    header: 'HARGA',
    cell: (info) => (
      <p className="font-semibold">{stringToCurrency(info.getValue())}</p>
    ),
    sortDescFirst: true,
  }),
  columnHelper.accessor('priceChanges.day', {
    header: '24 JAM',
    cell: (info) => <Percentage value={info.getValue()} />,
    sortDescFirst: true,
  }),
  columnHelper.accessor('priceChanges.week', {
    header: '1 MGG',
    cell: (info) => <Percentage value={info.getValue()} />,
    sortDescFirst: true,
  }),
  columnHelper.accessor('priceChanges.month', {
    header: '1 BLN',
    cell: (info) => <Percentage value={info.getValue()} />,
    sortDescFirst: true,
  }),
  columnHelper.accessor('priceChanges.year', {
    header: '1 THN',
    cell: (info) => <Percentage value={info.getValue()} />,
    sortDescFirst: true,
  }),
]

const Table = ({ data }: ITableProps) => {
  const [sorting, setSorting] = useState<SortingState>([])

  const tableLength = data.length

  const table = useReactTable({
    data,
    columns,
    state: {
      sorting,
    },
    onSortingChange: setSorting,
    getCoreRowModel: getCoreRowModel(),
    getSortedRowModel: getSortedRowModel(),
  })

  return (
    <>
      <table className="w-full overflow-x-auto border border-separate rounded-lg border-slate-300 border-spacing-0">
        <thead className=" text-slate-400">
          {table.getHeaderGroups().map((headerGroup) => (
            <tr key={headerGroup.id}>
              {headerGroup.headers.map((header) => (
                <th
                  key={header.id}
                  className="p-5 border-b first-of-type:pl-[75px] text-left"
                >
                  {!header.isPlaceholder && (
                    <div
                      className={`flex items-center gap-4 ${
                        header.column.getCanSort()
                          ? 'cursor-pointer select-none'
                          : ''
                      }`}
                      onClick={header.column.getToggleSortingHandler()}
                    >
                      {flexRender(
                        header.column.columnDef.header,
                        header.getContext()
                      )}
                      {header.column.getCanSort() && (
                        <Sorting type={header.column.getIsSorted() as any} />
                      )}
                    </div>
                  )}
                </th>
              ))}
            </tr>
          ))}
        </thead>
        <tbody>
          {table.getRowModel().rows.map((row) => {
            const isLastIndex = tableLength - 1 === row.index
            return (
              <tr key={row.id}>
                {row.getVisibleCells().map((cell) => (
                  <td
                    key={cell.id}
                    className={`p-5 ${
                      isLastIndex ? 'border-none' : 'border-b'
                    }`}
                  >
                    {flexRender(cell.column.columnDef.cell, cell.getContext())}
                  </td>
                ))}
              </tr>
            )
          })}
        </tbody>
      </table>
    </>
  )
}

export default Table
