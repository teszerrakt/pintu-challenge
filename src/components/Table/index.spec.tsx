import { render, screen } from '@testing-library/react'
import { MOCK_LATEST_PRICE_DATA } from 'src/__mocks__/latest-price'
import Table from 'src/components/Table'

jest.mock('src/components/Icons/Logo')

const mockData = MOCK_LATEST_PRICE_DATA

describe('Table', () => {
  it('renders the table header correctly', () => {
    render(<Table data={mockData} />)

    // Assert the presence of the table headers
    const headerElements = screen.getAllByRole('columnheader')
    expect(headerElements.length).toBe(6)
    expect(headerElements[0]).toHaveTextContent('CRYPTO')
    expect(headerElements[1]).toHaveTextContent('HARGA')
    expect(headerElements[2]).toHaveTextContent('24 JAM')
    expect(headerElements[3]).toHaveTextContent('1 MGG')
    expect(headerElements[4]).toHaveTextContent('1 BLN')
    expect(headerElements[5]).toHaveTextContent('1 THN')
  })

  it('renders the table rows correctly', () => {
    render(<Table data={mockData} />)

    const rowElements = screen.getAllByRole('row')
    expect(rowElements.length).toBe(mockData.length + 1) // +1 for the header row

    const firstRow = rowElements[1]
    expect(firstRow).toMatchSnapshot()
  })
})
