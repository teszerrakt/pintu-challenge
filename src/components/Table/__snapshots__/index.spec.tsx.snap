// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders the table rows correctly 1`] = `
<tr>
  <td
    class="p-5 border-b"
  >
    <a
      class="flex items-center"
      href="undefined/market/BTC"
    >
      <div
        class="flex flex-col w-full ml-6 lg:justify-between lg:flex-row"
      >
        <p
          class="font-semibold"
        >
          Bitcoin
        </p>
        <p
          class="text-slate-400"
        >
          BTC
        </p>
      </div>
    </a>
  </td>
  <td
    class="p-5 border-b"
  >
    <p
      class="font-semibold "
    >
      Rp 458.103.826
    </p>
  </td>
  <td
    class="p-5 border-b"
  >
    <div
      class="flex gap-[5px] text-green-600 "
      data-testid="percentage"
    >
      <div
        class="movement-icon"
      >
        ▲
      </div>
      <p
        class="font-semibold"
      >
        0.66
        %
      </p>
    </div>
  </td>
  <td
    class="p-5 border-b"
  >
    <div
      class="flex gap-[5px] text-green-600 "
      data-testid="percentage"
    >
      <div
        class="movement-icon"
      >
        ▲
      </div>
      <p
        class="font-semibold"
      >
        2.26
        %
      </p>
    </div>
  </td>
  <td
    class="p-5 border-b"
  >
    <div
      class="flex gap-[5px] text-green-600 "
      data-testid="percentage"
    >
      <div
        class="movement-icon"
      >
        ▲
      </div>
      <p
        class="font-semibold"
      >
        10.47
        %
      </p>
    </div>
  </td>
  <td
    class="p-5 border-b"
  >
    <div
      class="flex gap-[5px] text-green-600 "
      data-testid="percentage"
    >
      <div
        class="movement-icon"
      >
        ▲
      </div>
      <p
        class="font-semibold"
      >
        51.33
        %
      </p>
    </div>
  </td>
</tr>
`;
